Add model for median cost, evaluate on hold out validation set and compute new metrics

sofiapinto opened the pull request:
---

# Description

In this PR we:
- Change hyperparameter tuning and cross validation scripts to integrate a model for the 50th percentile/q=0.5/median
- Evaluate on an additional hold out validation set (a sample from the newest quarter, which we don't process for the exclusion of outliers nor windsorize)
- Add new metrics related to the distance between median and upper/lower bounds

Closes  #51 #54 #55

# Instructions for Reviewer

Hey @caldwellst , thanks for reviewing this PR!

Could you please review the changes to `process_data_before_modelling()` in `asf_hp_cost_estimator_model/pipeline/data_processing/process_installations_data.py`

Might be useful to review these scripts as a whole rather than just the changes in this PR:
- `asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py`
- `asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py`
- `asf_hp_cost_estimator_model/utils/model_evaluation_utils.py`

Any comments on the hyperparameters (lines 121-140) in `asf_hp_cost_estimator_model/config/base.yaml` would be useful too. Do you think the upper model might be overfitting?

Would there be any other checks/metrics you think I should add?

In general, I have a list of improvements to make to the repo (some refactoring, some old files need to be deleted and some automation) but if you have any suggestions of improvements on these files in specific, do let me know.

# Checklist:

- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

sofiapinto committed:
process_installations_data: adding flag to identify when processing a validation set

sofiapinto committed:
utils: updating model evaludation utils with new metrics and move load and prepare function to this file

sofiapinto committed:
config: adding  model hyperparameters for median model

sofiapinto committed:
hyperparm tuning: updating script with using validation set and training the q0.5 model

sofiapinto committed:
cross validation: changes to reflect validation set, new metrics and q0.5 model

sofiapinto commented on asf_hp_cost_estimator_model/config/base.yaml:
```diff
@@ -132,3 +132,9 @@ hyper_parameters:
     min_samples_split: 1000
```
```
min_samples_leaf: 1
min_samples_split: 1000
```
might this mean that the model is overfitting?

caldwellst replied:
I would be interested in what the actual distribution of leaves look like in terms of `n` samples. I doubt it's overfitting, rather that there was no need to tune `min_samples_leaf` because you're only splitting on nodes of 1,000+ samples. So, unless that somehow is going to split 999 and 1, it's very likely that your leaves will have sufficient data.

caldwellst replied:
```
min_samples_leaf: 100
min_samples_split: 10
```

Similarly, your `min_samples_split` is essentially redundant here because you require all leaves to have at minimum 100 samples, so you could never split a node of 10 anyway, you'd need at least 200.

I'm pretty confident that the rules set for these models should be doing alright, though, and you can safely ignore the small numbers in both cases. Also, it's a random forest, so another buttress against overfitting. Since we're looking at unseen samples of data to test performance, we should have some certainty that we aren't overfitting, but can revisit these if we think we are.

caldwellst commented on .pre-commit-config.yaml:
```diff
@@ -1,6 +1,6 @@
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v5.0.0
+    rev: v6.0.0
```
Might add to an open issue to shift to the new pre-commit config.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -1,39 +1,32 @@
+- upper quantile
 
-After tuning the model, it evaluates the model on both training and test sets, logging various metrics such as mean pinball loss, coverage probability, and interval widths.
+After tuning the model, it evaluates the model on training, test sets and an unseen validation set of data from.
```
Sentence cuts off.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -61,126 +54,164 @@ def argparse_setup():
+    quantile: float,
+    param_grid: Dict[str, Any],
+    conf: Dict[str, Any],
+) -> Dict[str, Any]:
```
I'll let your lack of parameter documentation slide ;)

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -1,39 +1,32 @@
 from sklearn.model_selection import train_test_split
 from sklearn.metrics import make_scorer, mean_pinball_loss
 from sklearn.experimental import enable_halving_search_cv  # noqa
 from sklearn.model_selection import HalvingRandomSearchCV
```
This looks cool, and very efficient!

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -1,39 +1,32 @@
 """
-This script performs hyperparameter tuning for a Gradient Boosting Regressor model for lower and upper quantiles.
+This script performs hyperparameter tuning for a Gradient Boosting Regressor models for three different quantiles:
```
Can we remove redundant files from the repo? Seems a lot of the other scripts are no longer used in the new modeling, and worth keeping the repo clean. You can use releases if you want in order to make the old state findable again, but wouldn't prioritise that.

caldwellst replied:
For instance, was noticing `hyperparameter_tuning.py` and `params_to_tune.py`. What else can be dropped?

sofiapinto replied:
Yes, I totally agree - as I mentioned in the summary above, it's on my list to remove the old files. These files were never part of what's in production now, so I can just delete them. The one that was previously in production lives in an old repo, so i don't think we need releases for now. #46

caldwellst replied:
Yeah sorry, missed that summary of yours, doh!!

caldwellst commented on asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:
```diff
@@ -7,134 +7,218 @@
+    """
+
+    logging.info("Loading and processing data...")
+    mcs_epc_data = get_enhanced_installations_data()
```
I know this is outside scope of this review, but is this essentially depending on us manually changing the config every time there is new installations data? If we want to always be using the latest available, you could change `get_enhanced_installations_data()` to automatically pull in the latest available based on parsing filename dates in the bucket. If valuable, would add to the backlog.

sofiapinto replied:
Yep, it was already in the backlog :)

caldwellst commented on asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:
```diff
@@ -7,134 +7,218 @@
+    logging.info("Loading and processing data...")
+    mcs_epc_data = get_enhanced_installations_data()
+    cpi_df = get_cpi_data()
+    postcodes_data = get_postcodes_data()
```
Might be worth standardising postcode infilling and data in a medallion structure if we use again in another project, like HPS. For instance, if we ever want to link our data out to postcodes. Again, I know not in the scope of this PR, just flagging!

sofiapinto replied:
Great idea - will add to our backlog.

caldwellst commented on asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:
```diff
@@ -7,134 +7,218 @@
+        postcodes_data=postcodes_data,
+        cpi_quarterly_df=cpi_quarterly_df,
+        min_date=config["min_date"],
+        processing_validation_set=True,
```
Aha, I see what's happening with `processing_validation_set` now. So essentially you aren't windsorising or doing any of the processing on another variables either, that you do to the training data prior to running. Maybe a more logical parameter value here is something like `remove_samples` or `remove_outliers` or something, and only when `remove_samples` is `True` do you run `remove_samples_exclusion_criteria()` (and it's only then that `winsorise` is actually relevant as a parameter).

caldwellst commented on asf_hp_cost_estimator_model/pipeline/data_processing/process_installations_data.py:
```diff
@@ -637,6 +637,7 @@ def process_data_before_modelling(
     winsorise: str = config["winsorise_outliers"],
     min_date: dict = config["min_date"],
     rooms_as_categorical: bool = False,
+    processing_validation_set: bool = False,
```
This should come before `winsorise`, given `winsorise` isn't used if this is `True`.

caldwellst replied:
Also, would make it clearer what this parameter actually does. Have mentioned this elsewhere, but really what you're doing is removing samples that are considered outliers or otherwise should be excluded for training. Make that explicit in the parameter. Currently you're saying for what datasets you would want to process or not, but that is more confusing to interpret.

caldwellst commented on asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:
```diff
@@ -7,134 +7,218 @@
 
-    mae_upper = mean_absolute_error(y, y_pred_upper)
-    mae_lower = mean_absolute_error(y, y_pred_lower)
+    features = config["numeric_features"] + config["categorical_features"]
```
Not critical, but some config validation ensuring that these exclude `target_feature`.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -61,126 +54,164 @@ def argparse_setup():
+    y_pred_upper = models["upper"].predict(X_data)
+
+    # Compute and log metrics
+    compute_metrics(
```
So you're just running this for the side effects of generating the plots and logging the metrics I guess? Worth adding a return statement here in case you want to run and assign them to memory? Or save them to S3, particularly as the log statements are not structured.

sofiapinto replied:
These are saved later for the final model. I don't it's worth saving these.. We could save the ones that come from CV though?

caldwellst replied:
Disagree about worth saving these, because as just flagged in the other PR, I think saving them here would make for simpler code structure for checking performance drift!

sofiapinto replied:
Yep - i was proposing saving the results in the cross validation script, but since we're removing that script, I'll save these results now.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -61,126 +54,164 @@ def argparse_setup():
+    y_train: pd.Series,
+    quantile: float,
+    param_grid: Dict[str, Any],
+    conf: Dict[str, Any],
```
Don't pass this is as a parameter, it's merely taking in the imported config above.

caldwellst replied:
Similarly, if `param_grid` was part of the config, then this simply just tunes a model of data for a specific quantile.

sofiapinto replied:
wow.. i don't remember writing this! must have been a copilot suggestion I accidentally accepted.

sofiapinto replied:
agree with param_grid going to base.yaml. will move!

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -61,126 +54,164 @@ def argparse_setup():
-    model_upper = GradientBoostingRegressor(
-        loss="quantile", alpha=upper_quantile, **search_uq.best_params_
+    # Define the parameter grid for hyperparameter tuning
+    param_grid = dict(
```
Could consider putting this in the config since you're globally defining the hyperparameters to tune, and then passing through nested functions. Makes them busier than necessary when actually we just need to call the param grid from the config once in `tune_model()` I think!

sofiapinto replied:
yep, totally agree. will do

caldwellst commented on asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:
```diff
@@ -7,134 +7,218 @@
+    postcodes_data = get_postcodes_data()
+
+    cpi_quarterly_df = get_df_quarterly_cpi_with_adjustment_factors(
+        ref_year=config["cpi_data"]["cpi_reference_year"],
```
Similarly, if you were thinking of making this fully self-maintained, is there no way to automate the reference year based on the current date, what's in the CPI dataset, etc.? Might be worth an addition to the backlog of what will make the cost estimator fully self functioning.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -61,126 +54,164 @@ def argparse_setup():
+        final_models[name] = model
+
+    # Evaluate models on all three datasets
+    evaluate_and_log_metrics(
```
Seems that this is a manual step to update the hyperparameters in the config.yaml file based on logging here. Just as easily you could automatically change those values using the `yaml` library. Again, if you wanted to re-run this automatically, might be a step for the backlog.

sofiapinto replied:
Agree! :)

caldwellst commented on asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:
```diff
@@ -7,134 +7,218 @@
 
```
Update the docstring at the top.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py:
```diff
@@ -58,159 +49,171 @@ def argparse_setup():
-    lower_quantile = args.lower_quantile
-    upper_quantile = args.upper_quantile
+def create_model(
+    quantile: float, model_params: Dict[str, Any], random_state: int
```
`random_state` slightly redundant given this is just passed originally from your config, could just remove and use the config internally.

caldwellst commented on asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py:
```diff
@@ -10,27 +10,18 @@
 
 # package imports
-import numpy as np
+import pandas as pd
```
Won't go through this in detail, I think is totally unnecessary!

sofiapinto replied:
I don't think I understand. What's totally unnecessary? We're importing pandas and using it for typehints.

sofiapinto replied:
oh wait, you meant the whole cross_validation.py script?

caldwellst replied:
Sorry, meant the whole script!

caldwellst commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -61,126 +54,164 @@ def argparse_setup():
-        min_samples_leaf=[1, 5, 10, 100, 1000],
-        min_samples_split=[2, 10, 50, 100, 1000],
+    # Prepare validation set
+    X_val = validation_set[features]
```
Okay, I think there is a lot going on here. For one, I think the terminology is slightly confusing. Typically we think of a 3-fold split of our data train --> validation --> test. Here, though, you're already doing cross validation, so technically you don't need a separate "validation" dataset per se. In fact, what I think what you define as `X_test` and `y_test` is a 20% hold out test set on random data, and what you define as `X_val` and y_`val` is a hold out test set on the latest data only.

I think it's definitely good to do, but wanted the language to be clear. Because what you're doing is doing training and validation as you perform the grid search, then testing performance on two hold out test sets. The next step to run everything in `cross_validation.py` is completely unnecessary. You have already done all the tests necessary to align on the proper hyper parameters and check how performance is. All that is left is training a final model using those parameters on the entire dataset.

sofiapinto replied:
I think this one might be good to discuss more in person. Have added a quick call tomorrow, hope it works.

caldwellst reviewed (changes requested):
Hello! Some general comments fall under these categories:

- Places where automation could happen. Not making value statement on whether we should do that, but could go into the potential backlog.
- Some documentation comments. Critically, make sure the `README.md` is updated.
- A few recommendations on function structure.
- Cleanup! There is a lot of files in there that seem to be from the old model. We should be keeping the repo clean and tidy, and don't see a reason to maintain them in the repo.

The big one though is I think the `cross_validation.py` script is entirely unnecessary. All you need at the end of the hyperparameter tuning and testing on the holdout test set is to retrain a model on the entire dataset with those parameters. Details in the review comment, but happy to discuss further as necessary.

sofiapinto committed:
move param grid to base.yaml

sofiapinto committed:
model_evalution_utils.py: improvements to documentation

sofiapinto committed:
process_installations_data.py: improving function docs

sofiapinto committed:
tune_hyperparameters.py: changes requested by Seth

sofiapinto committed:
removing uncessary model evaluation files

sofiapinto committed:
removing unecessary/old hyperparameter tuning files

sofiapinto committed:
removing old model training files

sofiapinto committed:
adding a test param to tune_hyperparameters.py

sofiapinto committed:
changes to README.md following changes to pipeline scripts

sofiapinto commented:
Hey @caldwellst thank you for the the review - it was so useful! Hey @crispy-wonton, thank you for agreeing to do a final review of this.

Following Seth's review I have:
- Have made changes to the `README.md`
- Took note of places where automation could happen (as issues!)
- Remove the old files that we don't need anymore: all the files in model_evaluation, some files across all of our pipeline folders. The main important change since the beginning of this review is that the cross validation script isn't needed, as it's inbuilt into `HalvingRandomSearchCV()` which we're using to tune hyperparameters.
- Fixes/improvements to documentation (the docstrings at the top of the files + function documentation and typehints). Have also improved the documentation of the `process_data_before_modelling()` function (from `process_installations_data.py`) but I think we need to look at this function separately to simplify it and get rid of things we don't use anymore.  
- Besides logging results we're now also saving them to S3 so that we can analyse drift (separate PR has the drift analysis code)
- Moved `param_grid` to the config file
- I implemented automatically saving the hyperparameters to the config file, and it's working. However, it's removing all the docs from the config file and reorganising the config file, so i've left this commented for now and will take a look at it in a separate issue.
- Improved the hyperparameter script as a whole: we're now first creating a hold out test set with 20% of random samples from the newest quarter. With the remaining data, we're splitting it into a training and test sets (this test set is a random 20% of all data that is left after creating the previous hold out test set). The training set is then used to do hyperparameter tunning and cross validation. The two tests sets are both hold out tests sets and don't go through outlier removal/windsorisation so that we can check how the model is doing with "real life data". Metrics are saved to S3. Metrics are computed on the training and both tests sets. Additionally the average CV results are also stored.


@crispy-wonton the main task is re-reviewing this file `asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py` and checking it runs with:
`python asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py --lower_quantile 0.1 --upper_quantile 0.9 --test False`

Please pay special attention to the logic to make sure I haven't made any mistake. Please also leave any suggestions for changes to names of variables/how the code is organised (although at this point, this is secondary).

You'll see loads of diffs and loads of files, but that's mainly because I removed lots of different files from the pipeline.

Thanks so much in advance!

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -1,24 +1,39 @@
 
-Usage:
-python asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py --lower_quantile 0.1 --upper_quantile 0.9
+After tuning the models, they are evaluated on the training, test set and and most recent quarter hold-out test set.
```
```suggestion
After tuning the models, they are evaluated on the training, test set and most recent quarter hold-out test set.
```

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -1,24 +1,39 @@
-Usage:
-python asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py --lower_quantile 0.1 --upper_quantile 0.9
+After tuning the models, they are evaluated on the training, test set and and most recent quarter hold-out test set.
+The script loggs various metrics such as mean pinball loss, coverage probability, and interval widths.
```
```suggestion
The script logs various metrics such as mean pinball loss, coverage probability, and interval widths.
```

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -1,24 +1,39 @@
+
+The best hyperparameters and evaluation metrics are then saved as CSV files to S3.
+
+This script can be run from the command line, allowing for custom quantiles to be specified (and whether to run in test mode):
```
You could add here what you mean by 'test mode'

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+        )
+
+    # Update config with best hyperparameters
+    # with open(os.path.join(PROJECT_DIR,"asf_hp_cost_estimator_model/config/base.yaml"), "w") as f:
```
This might be helpful here: https://www.geeksforgeeks.org/python/how-to-append-data-to-yaml-file-using-python/

particularly the first method, Using YAML Python Module, because it preserves the integrity of the existing yaml file content.

sofiapinto replied:
thank you, I'll open a separate issue to tackle this!

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+    search = HalvingRandomSearchCV(
+        gbr,
+        param_grid,
+        resource="n_estimators",
+        max_resources=500,
+        min_resources=50,
```
For my understanding, how did you choose these parameters? I don't have a good sense of how to select the resources beyond the defaults.

sofiapinto replied:
In this case the min and max resources relate to the number of estimators. When we were using grid search before, the number of estimators tested was between 50 and 500, so we kept it the same here. Typically the value coming out in our tests is 450. If it was consistently 500 or 50 then I would have expanded the range of values.

crispy-wonton replied:
Cool that makes sense thank you!

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+        loss="quantile", alpha=quantile, random_state=config["random_state"]
+    )
+
+    search = HalvingRandomSearchCV(
```
As this now acts as your model cross-validation, you could explicitly state the strategy or k-folds here. The default is `cv=5` folds - you could add this param explicitly or annotate.

sofiapinto replied:
This is a good idea, explicitly stating the k-folds here and explicitly annotating this is doing CV.

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+        param_grid (Dict[str, List[Union[int, float]]]): Grid of hyperparameters to search over.
+
+    Returns:
+        Dict[str, Union[int, float]]: Best hyperparameters found during the search.
```
Looks like it's returning the Random Search object rather than the best parameters specifically.

sofiapinto replied:
yep, good flag, I changed the return and forgot to change the docstring!

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+    # Identifying first date of newest quarter to create a hold-out validation set from the most recent data
+    valid_dates = mcs_epc_data[mcs_epc_data["commission_date"] < datetime.today()]
+    max_date = valid_dates["commission_date"].max()
+    first_day_of_quarter = max_date.to_period("Q").start_time
```
clever!

crispy-wonton commented:
Fyi I had issues with the set up before being able to run the script.
I deleted my existing conda env for this project to re-create it and there were the following issues and fixes:
1. The `make install` command was hanging, I looked at the logs which showed the terminal was waiting for a response to update packages. I needed to enter 'y' and press enter in my terminal to proceed.
2. `make install` didn't install requirements after running to completion without error, so I had to run `pip install -r requirements.txt` after set up.
3. When trying to run the script, I got a `ModuleNotFoundError: No module named 'asf_hp_cost_estimator_model'`. I set a temporary fix in the script by importing `sys`, and then `sys.path.append('my/local/path')`. This is a recurring issue for me in this repository and I can't figure out the underlying issue.
4. I then tried to run the script again a few times and got another few `ModuleNotFoundError`: 

- `requests`
- `openpyxl`

I pip installed the packages as they came up. It looks like they are missing from `requirements.txt`. Not sure if you're using a different setup methodology now so maybe this is already fixed but I didn't see it in the README.

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+
+
+def load_and_prepare_data(
+    model_data: pd.DataFrame,
+    test_set: pd.DataFrame,
+    new_quarter_hold_out_set: pd.DataFrame,
```
These are missing from docstring

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+    )
+    mcs_epc_data = mcs_epc_data.drop(new_quarter_hold_out_set.index)
+
+    return mcs_epc_data, new_quarter_hold_out_set
```
Could be nice to print out here the size of the hold out set and remaining data set.

sofiapinto replied:
Yes, those values are logged later in the pipeline :)

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+    """Loads and preprocesses all necessary data prior to hyperparameter tuning or cross validation.
+
+    Returns:
+        Tuple[pd.DataFrame, pd.DataFrame, List[str]]: Processed model data, validation set, and list of feature names.
```
The return docstring needs updating

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+    return search
+
+
+def split_data(mcs_epc_data: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame]:
```
would suggest renaming e.g. something like `extract_hold_out_data` so that the result of the function is more immediately clear when calling it.

sofiapinto replied:
good shout, will do

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
-    )
+        # Mean cross-validated score (mean pinball loss) of the best estimator
+        cv_scores_name = f"mean_pinball_loss_{name}"
+        cv_scores_name += "_q" if name != "median" else ""
```
do you need the `else` here?

sofiapinto replied:
i believe you get a SyntaxError otherwise

crispy-wonton replied:
oh sad 😞

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
+        # Mean cross-validated score (mean pinball loss) of the best estimator
+        cv_scores_name = f"mean_pinball_loss_{name}"
+        cv_scores_name += "_q" if name != "median" else ""
+        cv_scores[cv_scores_name] = -tuned_model.best_score_
```
How come you have put the minus sign in front of `tuned_model`? I see in the `sklearn` `make_scorer` [docs](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html) that the scorer object will sign-flip the outcome of the `score_func` if using a loss function...is it related to this?

If so, how should we interpret the scores in the `cv_scores` dict? Does higher score mean a lower loss?

sofiapinto replied:
Hey - we've discussed this yesterday but make scorer flips the sign, but in the outputs we want to take a look at the positive values. 

(When the sign isn't flipped) The mean pinball loss is a value higher than 0, which we want to minimise. A higher values means being further way from the correct prediction.

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:
```diff
@@ -58,129 +72,287 @@ def argparse_setup():
-    logging.info("Best params for upper percentile model: ")
-    pprint.pprint(search_uq.best_params_)
+    # Train final models on the full training set using the best parameters
+    logging.info("Training final models with best hyperparameters...")
+    final_models = {}
+    for name, params in best_params.items():
+        quantile = models_to_tune[name]
+        model = GradientBoostingRegressor(
+            loss="quantile",
+            alpha=quantile,
+            random_state=config["random_state"],
+            **params,
+        )
+        model.fit(X, y)
+        final_models[name] = model
```
I think it's possible you don't need this code. I'm reading the `HalvingRandomSearchCV` [docs](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.HalvingRandomSearchCV.html) and there's a param called `refit` which defaults to `True`. The description: "Refit an estimator using the best found parameters on the whole dataset."

I think you can access these using `tuned_model.best_estimator_` for each tuned model. You can then use `predict` directly on the `best_estimator_` attribute.

sofiapinto replied:
This is a good flag. I had seen that but didn't have time to explore it, so I'll open a separate issue for this.

crispy-wonton reviewed (commented):
I've reviewed `tune_hyperparameters.py`.

I left a separate comment in the main thread about issues I had setting up and how I fixed them. After that, the script ran successfully all the way through and the parameters and metrics are saved successfully to s3.

I didn't find any issues with the logic. There are some suggestions for improving the docstrings. I also had a question about the negative pinball loss score which we discussed offline but I've left the comment as you were planning to test out `greater_is_better` with True and False.

I also left a suggestion about accessing the `best_estimator_` attribute directly rather than retraining the model. 

I'll have a quick look over the other files you mentioned separately after publishing this review.

crispy-wonton commented on asf_hp_cost_estimator_model/pipeline/data_processing/process_installations_data.py:
```diff
@@ -671,11 +681,12 @@ def process_data_before_modelling(
-        exclusion_criteria_dict=exclusion_criteria_dict,
-        winsorise=winsorise,
-    )
+    if remove_or_winsorise_samples:
+        enhanced_installations_data = remove_samples_exclusion_criteria(
+            mcs_epc_data=enhanced_installations_data,
+            exclusion_criteria_dict=exclusion_criteria_dict,
+            winsorise=winsorise,
+        )
```
Does this work if you only do one or the other? E.g. if you want to winsorise but provide no exclusion criteria, will it work or error out and vice versa?

sofiapinto replied:
Yes, it works!

The exclusion criteria covers a lot of variables (floor area, number of rooms, etc - not just cost). You can provide no exclusion criteria but still winsorise or winsorise and not exclude any data. You can also exclude data and winsorise. 

For cost this is what happens: we only winsorise if there's no exclusion criteria for cost provided. When winsorising you can choose to winsorise both ends, only one or none!

```
if (
        ("ajusted_cost_upper_bound" not in exclusion_criteria_dict)
        and ("cost_upper_bound" not in exclusion_criteria_dict)
        and ("adjusted_cost_lower_bound" not in exclusion_criteria_dict)
        and ("cost_lower_bound" not in exclusion_criteria_dict)
    ):
        filtered_data = remove_or_winsorise_cost_outliers_per_group(
            data=filtered_data,
            target_feature=config["target_feature"],
            winsorise=winsorise,
        )
```

crispy-wonton reviewed (commented):
One additional comment on the code in `process_installations_data.py`.

sofiapinto committed:
Update asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py

Co-authored-by: crispy-wonton <104171770+crispy-wonton@users.noreply.github.com>

sofiapinto committed:
Update asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py

Co-authored-by: crispy-wonton <104171770+crispy-wonton@users.noreply.github.com>

sofiapinto committed:
roisin's suggestions

sofiapinto commented:
> Fyi I had issues with the set up before being able to run the script. I deleted my existing conda env for this project to re-create it and there were the following issues and fixes:
> 
> 1. The `make install` command was hanging, I looked at the logs which showed the terminal was waiting for a response to update packages. I needed to enter 'y' and press enter in my terminal to proceed.
> 2. `make install` didn't install requirements after running to completion without error, so I had to run `pip install -r requirements.txt` after set up.
> 3. When trying to run the script, I got a `ModuleNotFoundError: No module named 'asf_hp_cost_estimator_model'`. I set a temporary fix in the script by importing `sys`, and then `sys.path.append('my/local/path')`. This is a recurring issue for me in this repository and I can't figure out the underlying issue.
> 4. I then tried to run the script again a few times and got another few `ModuleNotFoundError`:
> 
> * `requests`
> * `openpyxl`
> 
> I pip installed the packages as they came up. It looks like they are missing from `requirements.txt`. Not sure if you're using a different setup methodology now so maybe this is already fixed but I didn't see it in the README.

@crispy-wonton thank you for the detailed message. I haven't been able to reproduce the issue on my end, so I'll open a new issue  to solve this separately. I'll move us to use uv which will hopefully sort this issue.

sofiapinto committed:
model_evaluation_utils.py: saving histograms with dataset name

sofiapinto committed:
save histograms

sofiapinto committed:
add random state when sampling
